### DISCUSSION TITLE
Explore options to integrate a live streaming solution

### DISCUSSION THREAD

**Discussion by mgogoulos** [mediacms-io/mediacms#638] [Feature Requests]
**Describe the feature you'd like**
Explore which of the following projects could be integrated to MediaCMS to provide live streaming too

- [ ] https://owncast.online/
- [ ] https://openstreamingplatform.com/
- [ ] https://github.com/livekit/livekit-server

**Comment by MyBlood** [comment]
> **Describe the feature you'd like** Explore which of the following projects could be integrated to MediaCMS to provide live streaming too
> 
> * [x]  https://owncast.online/
> * [ ]  https://openstreamingplatform.com/
> * [ ]  https://github.com/livekit/livekit-server

**Livekit** and **Owncast** look great. But alas, from a developer's point of view, I can't say which is better.

**Comment by KyleMaas** [comment]
I'd be really interested in seeing this feature as well.

**Comment by KyleMaas** [comment]
OpenStreamingPlatform looks to be fairly easy to integrate using the OAuth2 support and Webhooks.  If I understand it correctly, the MediaCMS Django instance would need to be set up as an OAuth2 provider so logins could be used across both platforms:

https://django-oauth-toolkit.readthedocs.io/en/latest/tutorial/tutorial_01.html

From there, MediaCMS would need to implement a Webhook endpoint for OSP to catch new videos, comments, etc. and add them as uploaded videos:

https://open-streaming-platform.readthedocs.io/en/latest/usage/webhooks.html

**Reply by KyleMaas** [reply]
Tried to get this working on the same server as MediaCMS.  I wouldn't recommend that.  It requires several changes to the default OSP configuration and some of the libraries it requires do not like to work with Ubuntu 20.04 as suggested in the MediaCMS install docs.  If anything, make it easier on yourself and install to a separate server.  I still haven't actually gotten it to work.

**Reply by KyleMaas** [reply]
Well, it appears that installing OSP broke transcoding within MediaCMS.  I still haven't figured out why, but I would highly recommend against trying to install OSP on the same server as MediaCMS.  They do not appear to be compatible with each other.

**Comment by KyleMaas** [comment]
Owncast can also be set up with a Webhook endpoint:

https://owncast.online/thirdparty/webhooks/

However, as far as I can tell, it only supports one streaming user - the Owncast administrator.  This could work for situations where MediaCMS only has one poster (internal corporate environment, for example) but may not work as well for a publicly-accessible instance where more than one user can post media to the site and would want to be able to live stream.

**Reply by KyleMaas** [reply]
I have spent a lot of time trying to get Owncast to work, and thus far my efforts have yielded zero fruit.  It installs, it runs, but then when I try to stream to it, it doesn't generate the m3u8 or segment files for the browser to actually use.  I'm not ready to give up on it yet, because if I can get it to work it could potentially be a really easy install.  But so far no luck.

**Comment by KyleMaas** [comment]
LiveKit looks to be much lower-level as far as implementation.  Does not include any functionality for chats or stream recording as far as I can tell, and user authentication is at a layer above the scope of the project.  It can be controlled using server APIs, and there is a Python SDK project available for it:

https://github.com/tradablebits/livekit-server-sdk-python

Also uses Webhooks for communicating back:

https://docs.livekit.io/server/webhooks/

**Comment by mgogoulos** [comment]
Hi @KyleMaas , since you're actively exploring this, did you have any chance to look at Jitsi or BigBlueButton? I've been using these systems for years and are very stable, an integration with any of them would be something to explore

**Reply by KyleMaas** [reply]
I've used Jitsi before and did not find it stable enough to consider for this.  But I'd be willing to look into it.  It looks like we'd want to aim for Jitsi Meet integration:

https://github.com/jitsi/jitsi-meet

Have not looked into BigBlueButton for feasibility but can investigate.

Have been trying to get Owncast to work but have had trouble getting it to run properly.

**Comment by mgogoulos** [comment]
I've been using both (Jitsi and BigBlueButton) on different projects recently and had a very positive experience! I think it definitely worths to explore installing them (should be very easy) and whether they provide a quick way to embed inside another web application (such as MediaCMS in this case). As far as I remember regarding Owncast, you still need to use more tools (eg OBS) to take care of the streaming

**Reply by mgogoulos** [reply]
These are some notes I had prepared related to Owncast - adding here just in case it provides any help


install OBS studio
Install and start Owncast - on Ubuntu/Linux, and probably other systems too, this is easy through:

```
curl -s https://owncast.online/install.sh | bash
cd owncast && ./owncast
```

This gives an rtmp URL and a key (eg rtmp://localhost:1935/live and abc123)

On OBS --> Settings --> Stream, select "custom" and provide URL+key. Then you can start streaming, for example add `camera`, or `video file`, OBS supports many options

Owncast has also an admin where you can see how many users are online. Plus a section for messages
Also related: https://owncast.online/docs/embed/

**Reply by KyleMaas** [reply]
Yeah, that's roughly what I was doing.  You can track my efforts at:

https://github.com/owncast/owncast/issues/2399

Would really kind of like to get that working because it's just so simple if it actually worked.

**Comment by dcwestra** [comment]
I know this is dated discussion, but I found a different solution with tools that have developed in the last 2.5 years. Background: I currently use owncast for live streaming and mediacms for video archiving that both have feeds embedded into a website that I host. I wanted to find a way to automatically record and upload the livestream so that it requires as little human effort as possible.

n8n has been making the rounds recently in the tech youtuber space and I decided to see if this could work - and it did!

Here is my n8n workflow:
1) Webhook trigger - n8n gives the webhook url that you can put into owncast
    a) Set the trigger to POST
    b) Set the owncast webhook to trigger on both stream start and stream stop
2) SSH block - point to a host with ffmpeg installe. Make sure directories exist and set to execute command
```
ffmpeg -hwaccel nvdec -i https://your-owncast-fqdn-or-ip:port/hls/stream.m3u8 -c:v h264_nvenc -preset fast -c:a aac -b:a 128k recordings/stream-name-$(date +%F).mp4
```
3) Use wait block - set resume to "on webhook call", it will look for the same webhook url from step 1.
4) SSH block - point to the same host and use this cli command to upload to mediacms via API:
```
curl -X POST https://mediacms.url/api/v1/media -u 'username:password' -F "title=Video Title" -F "description=Auto-uploaded from n8n webhook" -F "media_file=path/to/your/recording.mp4"
```
